File: backend/wallets/serializers/transaction.py
from rest_framework import serializers

from wallets.models import Transaction


class TransactionSerializer(serializers.ModelSerializer):
    asset_symbol = serializers.CharField(source="asset.symbol", read_only=True)
    asset_name = serializers.CharField(source="asset.name", read_only=True)
    wallet_address = serializers.CharField(source="wallet.address", read_only=True)

    class Meta:
        model = Transaction
        fields = (
            "uuid",
            "tx_hash",
            "chain",
            "from_address",
            "to_address",
            "asset",
            "asset_symbol",
            "asset_name",
            "amount",
            "market_value",
            "block_timestamp",
            "block_number",
            "status",
            "transaction_fee_estimated",
            "transaction_fee",
            "wallet",
            "wallet_address",
            "created_at",
        )
        read_only_fields = fields

File: packages/packages/shared-types/src/domain/transaction.ts
import type { BaseEntity } from '../common';
import type { BaseQueryParams, DateRangeParams } from '../api';

export interface Transaction extends BaseEntity {
  txHash: string;
  chain: string;
  fromAddress: string;
  toAddress: string | null;
  asset: string;
  assetSymbol?: string;
  assetName?: string;
  amount: string;
  marketValue?: string | null;
  blockTimestamp: string;
  blockNumber?: number;
  status?: string;
  transactionFeeEstimated?: string;
  transactionFee?: string;
  wallet: string;
  walletAddress?: string;

  orderType?: 'BUY' | 'SELL' | 'ISSUANCE' | 'MINT' | 'PAYMENT_SENT' | 'PAYMENT_RECEIVED';
  pricePerShare?: string;
  paymentToken?: string;
  transactionType?: string;
  issuanceType?: string;
}

export interface TransactionQueryParams extends BaseQueryParams, DateRangeParams {
  wallet?: string;
  chain?: string;
  asset?: string;
  direction?: 'incoming' | 'outgoing';
  min_amount?: number;
  max_amount?: number;
}
